[postgres-ha] unable to recover when primary pod is other than 0

<!--
If you are reporting a new issue, make sure that we do not have any duplicates
already open. You can ensure this by searching the issue list for this
repository. If there is a duplicate, please close your issue and add a comment
to the existing issue instead.

If you suspect your issue is a bug, please edit your issue description to
include the BUG REPORT INFORMATION shown below. If you fail to provide this
information within 7 days, we cannot debug your issue and we'll close it. We
will, however, reopen it if you later provide the information.
-------------------------------
    BUG REPORT INFORMATION
-------------------------------
Use the commands below to provide key information from your environment:
You do NOT have to include this information if this is a FEATURE REQUEST
-->

**Which chart**:

name: postgresql-ha
version: 1.4.8

**Description**

I found an interesting scenario where `postgres-ha` is unable to start/recover after update of helm chart with modified `replicaCount`.

It happens to me during failover tests of latest chart version. First I delete manually primary pod so that `repmgr` promote another replica to be the primary pod like `postgresql-ha-postgresql-1`.

:thinking: :thought_balloon: Thinking that we may need another replica that serves as a witness in such conditions,

OR being able to promote itself as primary if no other postgres pods are running.

This issue https://github.com/bitnami/charts/issues/1934 seems same behavior to me but it was fixed by bumping pg-pool image version.

Thanks for all the effort here. This chart could be splendid!

**Steps to reproduce the issue:**


1. Start `postgres-ha` chart with `2 replicas` using following `values.yaml`

```
---
  postgresql:
    username: postgres
    password: superuserpass
    database: app
    repmgrUsername: repmgr
    repmgrPassword: repmgrpass
    replicaCount: 2
  postgresqlImage:
    debug: true

  pgpool:
    adminPassword: pgpoolpass
    replicaCount: 2
```

2. Wait for pods to started

```
kubectl rollout status sts --namespace=postgresql-ha postgresql-ha-postgresql
```

3. Manually delete `postgresql-ha-postgresql-0` which has the primary role

```
kubectl delete po --namespace=postgresql-ha postgresql-ha-postgresql-0
```

4. Wait for deleted pod to be recreated

```
kubectl rollout status sts --namespace=postgresql-ha postgresql-ha-postgresql
```

5. Check which pod has the primary role using `repmgr`. I login to `postgresql-ha-postgresql-0` as user `postgres` which I manually created in order to run `repmgr` commands inside this non-root container.

```
postgres@postgresql-ha-postgresql-0:/$ repmgr cluster show -f /opt/bitnami/repmgr/conf/repmgr.conf --compact
 ID   | Name                       | Role    | Status    | Upstream                   | Location | Prio. | TLI
------+----------------------------+---------+-----------+----------------------------+----------+-------+-----
 1000 | postgresql-ha-postgresql-0 | standby |   running | postgresql-ha-postgresql-1 | default  | 100   | 2
 1001 | postgresql-ha-postgresql-1 | primary | * running |                            | default  | 100   | 2
```

6. Update helm chart to use `1 replica only`. Kubernetes first delete pod with index 1, then delete pod with index 0, and when it starts the new pod with index 0 it fails to recover forever with following message

```
[2020-03-11 12:06:18] [DETAIL] attempted to connect using:
user=repmgr password=repmgrpass connect_timeout=5 dbname=repmgr host=postgresql-ha-postgresql-1.postgresql-ha-postgresql-headless.postgresql-ha.svc.cluster.local port=5432 fallback_application_name=repmgr
[2020-03-11 12:06:18] [ERROR] unable connect to upstream node (ID: 1001), terminating
[2020-03-11 12:06:18] [HINT] upstream node must be running before repmgrd can start
```

At this moment, I think that repmgr on pod-0 is looking for pod-1 which was the last promoted primary, but since pod-1 does not exist, there is a missing logic for pod-0 to promote itself as primary.


**Describe the results you received:**

After scale out, the postgresql pod is continuously restarting without opportunity to promote itself as primary. Server is unreachable and I was unable to recover even using `existingClaim` or `pdb: true`

**Describe the results you expected:**

The postgresql server to remain in working condition with pod-0 as primary role.

**Additional information you deem important (e.g. issue happens only occasionally):**



**Version of Helm and Kubernetes**:

- Output of `helm version`:

```
Helm 2 using Rancher v2.3.5
```

- Output of `kubectl version`:

```
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.3", GitCommit:"06ad960bfd03b39c8310aaf92d1e7c12ce618213", GitTreeState:"clean", BuildDate:"2020-02-11T18:14:22Z", GoVersion:"go1.13.6", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.3", GitCommit:"06ad960bfd03b39c8310aaf92d1e7c12ce618213", GitTreeState:"clean", BuildDate:"2020-02-11T18:07:13Z", GoVersion:"go1.13.6", Compiler:"gc", Platform:"linux/amd64"}

```

--- a/bitnami/postgresql-ha/Chart.yaml
+++ b/bitnami/postgresql-ha/Chart.yaml
@@ -1,6 +1,6 @@
 apiVersion: v1
 name: postgresql-ha
-version: 3.1.1
+version: 3.2.0
 appVersion: 11.7.0
 description: Chart for PostgreSQL with HA architecture (using Replication Manager (repmgr) and Pgpool).
 keywords:

--- a/bitnami/postgresql-ha/templates/postgresql/hooks-scripts-configmap.yaml
+++ b/bitnami/postgresql-ha/templates/postgresql/hooks-scripts-configmap.yaml
@@ -1,0 +1,67 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ printf "%s-hooks-scripts" (include "postgresql-ha.postgresql" .) }}
+  labels: {{- include "postgresql-ha.labels" . | nindent 4 }}
+    app.kubernetes.io/component: postgresql
+data:
+  pre-stop.sh: |-
+    #!/bin/bash
+    set -o errexit
+    set -o pipefail
+    set -o nounset
+
+    # Debug section
+    exec 3>&1
+    exec 4>&2
+
+    # Load Libraries
+    . /opt/bitnami/scripts/liblog.sh
+    . /opt/bitnami/scripts/libpostgresql.sh
+    . /opt/bitnami/scripts/librepmgr.sh
+
+    # Auxiliar functions
+    is_new_primary_ready() {
+        return_value=1
+        currenty_primary_node="$(repmgr_get_primary_node)"
+        currenty_primary_host="$(echo $currenty_primary_node | awk '{print $1}')"
+
+        info "$currenty_primary_host != $REPMGR_NODE_NETWORK_NAME"
+        if [[ $(echo $currenty_primary_node | wc -w) -eq 2 ]] && [[ "$currenty_primary_host" != "$REPMGR_NODE_NETWORK_NAME" ]]; then
+            info "New primary detected, leaving the cluster..."
+            return_value=0
+        else
+            info "Waiting for a new primary to be available..."
+        fi
+        return $return_value
+    }
+
+    export MODULE="pre-stop-hook"
+
+    if [[ "${BITNAMI_DEBUG}" == "true" ]]; then
+        info "Bash debug is on"
+    else
+        info "Bash debug is off"
+        exec 1>/dev/null
+        exec 2>/dev/null
+    fi
+
+    # Load PostgreSQL & repmgr environment variables
+    eval "$(repmgr_env)"
+    eval "$(postgresql_env)"
+
+    postgresql_enable_nss_wrapper
+
+    # Prepare env vars for managing roles
+    primary_node="$(repmgr_get_primary_node)"
+    primary_host="$(echo $primary_node | awk '{print $1}')"
+
+    # Stop postgresql for graceful exit.
+    postgresql_stop
+
+    if [[ "$primary_host" == "$REPMGR_NODE_NETWORK_NAME" ]]; then
+        info "Primary node need to wait for a new primary node before leaving the cluster"
+        retry_while is_new_primary_ready 10 5
+    else
+        info "Standby node doesn't need to wait, leaving the cluster."
+    fi

--- a/bitnami/postgresql-ha/templates/postgresql/statefulset.yaml
+++ b/bitnami/postgresql-ha/templates/postgresql/statefulset.yaml
@@ -68,6 +68,11 @@ spec:
       {{- end }}
       containers:
         - name: postgresql
+          lifecycle:
+            preStop:
+              exec:
+                command:
+                  - /pre-stop.sh
           image: {{ include "postgresql-ha.postgresqlImage" . }}
           imagePullPolicy: {{ .Values.postgresqlImage.pullPolicy | quote }}
           {{- if .Values.postgresql.securityContext.enabled }}
@@ -205,6 +210,9 @@ spec:
             - name: data
               mountPath: {{ .Values.persistence.mountPath }}
             {{- end }}
+            - name: hooks-scripts
+              mountPath: /pre-stop.sh
+              subPath: pre-stop.sh
         {{- if .Values.metrics.enabled }}
         - name: metrics
           image: {{ include "postgresql-ha.metricsImage" . }}
@@ -264,6 +272,10 @@ spec:
             {{- end }}
         {{- end }}
       volumes:
+        - name: hooks-scripts
+          configMap:
+            name: {{ printf "%s-hooks-scripts" (include "postgresql-ha.postgresql" .) }}
+            defaultMode: 0755
         {{- if or (.Files.Glob "files/repmgr.conf") (.Files.Glob "files/postgresql.conf") (.Files.Glob "files/pg_hba.conf") .Values.postgresql.repmgrConfiguration .Values.postgresql.configuration .Values.postgresql.pgHbaConfiguration .Values.postgresql.configurationCM }}
         - name: postgresql-config
           configMap:

--- a/bitnami/postgresql-ha/values-production.yaml
+++ b/bitnami/postgresql-ha/values-production.yaml
@@ -244,7 +244,6 @@ postgresql:
   ## ref: https://github.com/bitnami/bitnami-docker-postgresql-repmgr#configuration-file
   ##
   # repmgrConfiguration: |-
-
   ## PostgreSQL configuration
   ## Specify runtime configuration parameters as a dict, using camelCase, e.g.
   ## {"sharedBuffers": "500MB"}
@@ -252,7 +251,6 @@ postgresql:
   ## ref: https://github.com/bitnami/bitnami-docker-postgresql-repmgr#configuration-file
   ##
   # configuration:
-
   ## PostgreSQL client authentication configuration
   ## Specify content for pg_hba.conf
   ## Default: do not create pg_hba.conf
@@ -263,24 +261,20 @@ postgresql:
   #   local all all trust
   #   host all all localhost trust
   #   host mydatabase mysuser 192.168.0.0/24 md5
-
   ## ConfigMap with PostgreSQL configuration
   ## NOTE: This will override repmgrConfiguration, configuration and pgHbaConfiguration
   ##
   # configurationCM:
-
   ## PostgreSQL extended configuration
   ## As above, but _appended_ to the main configuration
   ## Alternatively, you can put your *.conf under the files/conf.d/ directory
   ## ref: https://github.com/bitnami/bitnami-docker-postgresql-repmgr#allow-settings-to-be-loaded-from-files-other-than-the-default-postgresqlconf
   ##
   # extendedConf:
-
   ## ConfigMap with PostgreSQL extended configuration
   ## NOTE: This will override extendedConf
   ##
   # extendedConfCM:
-
   ## initdb scripts
   ## Specify dictionary of scripts to be run at first boot
   ## Alternatively, you can put your scripts under the files/docker-entrypoint-initdb.d directory
@@ -289,12 +283,10 @@ postgresql:
   #   my_init_script.sh: |
   #      #!/bin/sh
   #      echo "Do something."
-
   ## ConfigMap with scripts to be run at first boot
   ## NOTE: This will override initdbScripts
   ##
   # initdbScriptsCM:
-
   ## Secret with scripts to be run at first boot
   ## Note: can be used with initdbScriptsCM or initdbScripts
   ##
@@ -543,18 +535,18 @@ metrics:
     ## [Kube Prometheus Selector Label](https://github.com/bitnami/charts/tree/master/bitnami/prometheus-operator#exporters)
     selector:
       prometheus: kube-prometheus
-    
+
     ## RelabelConfigs to apply to samples before scraping
     ## ref: https://github.com/coreos/prometheus-operator/blob/master/Documentation/api.md#relabelconfig
     ## Value is evalued as a template
     ##
-    # relabelings: []
+    relabelings: []
 
     ## MetricRelabelConfigs to apply to samples before ingestion
     ## ref: https://github.com/coreos/prometheus-operator/blob/master/Documentation/api.md#relabelconfig
     ## Value is evalued as a template
     ##
-    # metricRelabelings: []
+    metricRelabelings: []
 
 ## Persistence paramaters
 ##

--- a/bitnami/postgresql-ha/values-production.yaml
+++ b/bitnami/postgresql-ha/values-production.yaml
@@ -29,7 +29,7 @@
 postgresqlImage:
   registry: docker.io
   repository: bitnami/postgresql-repmgr
-  tag: 11.7.0-debian-10-r93
+  tag: 11.7.0-debian-10-r96
   ## Specify a imagePullPolicy. Defaults to 'Always' if image tag is 'latest', else set to 'IfNotPresent'
   ## ref: http://kubernetes.io/docs/user-guide/images/#pre-pulling-images
   ##
@@ -90,7 +90,7 @@ volumePermissionsImage:
 metricsImage:
   registry: docker.io
   repository: bitnami/postgres-exporter
-  tag: 0.8.0-debian-10-r98
+  tag: 0.8.0-debian-10-r99
   ## Specify a imagePullPolicy. Defaults to 'Always' if image tag is 'latest', else set to 'IfNotPresent'
   ## ref: http://kubernetes.io/docs/user-guide/images/#pre-pulling-images
   ##

--- a/bitnami/postgresql-ha/values.yaml
+++ b/bitnami/postgresql-ha/values.yaml
@@ -533,28 +533,30 @@ metrics:
   ##
   serviceMonitor:
     enabled: false
+
     ## Specify a namespace if needed
     # namespace: monitoring
     # fallback to the prometheus default unless specified
     # interval: 10s
     # scrapeTimeout: 10s
+
     ## Defaults to what's used if you follow CoreOS [Prometheus Install Instructions](https://github.com/bitnami/charts/tree/master/bitnami/prometheus-operator#tldr)
     ## [Prometheus Selector Label](https://github.com/bitnami/charts/tree/master/bitnami/prometheus-operator#prometheus-operator-1)
     ## [Kube Prometheus Selector Label](https://github.com/bitnami/charts/tree/master/bitnami/prometheus-operator#exporters)
     selector:
       prometheus: kube-prometheus
-    
+
     ## RelabelConfigs to apply to samples before scraping
     ## ref: https://github.com/coreos/prometheus-operator/blob/master/Documentation/api.md#relabelconfig
     ## Value is evalued as a template
     ##
-    # relabelings: []
+    relabelings: []
 
     ## MetricRelabelConfigs to apply to samples before ingestion
     ## ref: https://github.com/coreos/prometheus-operator/blob/master/Documentation/api.md#relabelconfig
     ## Value is evalued as a template
     ##
-    # metricRelabelings: []
+    metricRelabelings: []
 
 ## Persistence paramaters
 ##

--- a/bitnami/postgresql-ha/values.yaml
+++ b/bitnami/postgresql-ha/values.yaml
@@ -29,7 +29,7 @@
 postgresqlImage:
   registry: docker.io
   repository: bitnami/postgresql-repmgr
-  tag: 11.7.0-debian-10-r93
+  tag: 11.7.0-debian-10-r96
   ## Specify a imagePullPolicy. Defaults to 'Always' if image tag is 'latest', else set to 'IfNotPresent'
   ## ref: http://kubernetes.io/docs/user-guide/images/#pre-pulling-images
   ##
@@ -90,7 +90,7 @@ volumePermissionsImage:
 metricsImage:
   registry: docker.io
   repository: bitnami/postgres-exporter
-  tag: 0.8.0-debian-10-r98
+  tag: 0.8.0-debian-10-r99
   ## Specify a imagePullPolicy. Defaults to 'Always' if image tag is 'latest', else set to 'IfNotPresent'
   ## ref: http://kubernetes.io/docs/user-guide/images/#pre-pulling-images
   ##

